restcomm/restcomm.interpreter/src/main/java/org/restcomm/connect/interpreter/NumberSelectorService.java

```diff
@@ -0,0 +1,325 @@
+    }
+
+    /**
+    * The main logic is:
+    * -Find a perfect match in DB using different formats.
+    * -If not matched, use available Regexes in the organization.
+    * -If not matched, try with the special * match.
+    *
+     * @param phone
+     * @param sourceOrganizationSid
+     * @param destinationOrganizationSid
+     * @return
+     */
+    public IncomingPhoneNumber searchNumber(String phone,
+            Sid sourceOrganizationSid, Sid destinationOrganizationSid) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("getMostOptimalIncomingPhoneNumber: " + phone
+                    + ",srcOrg:" + sourceOrganizationSid
+                    + ",destOrg:" + destinationOrganizationSid);
+        }
+        List<String> numberQueries = createPhoneQuery(phone);
+
+        IncomingPhoneNumber numberfound = findByNumber(numberQueries, sourceOrganizationSid, destinationOrganizationSid);
+        if (numberfound == null) {
+            //only use regex if perfect match didnt worked
+            if (destinationOrganizationSid != null &&
+                    destinationOrganizationSid.equals(sourceOrganizationSid)
+                    && phone.matches("[\\d,*,#,+]+")) {
+                //check regex if source and dest orgs are the same
+                //only use regex if org available
+                //check if there is a Regex match only if parameter is a String aka phone Number
+                numberfound = findByRegex(numberQueries, sourceOrganizationSid, destinationOrganizationSid);
+            }
+            if (numberfound == null) {
+                //if no regex match found, try with special star number in the end
+                numberfound = findSingleNumber("*", sourceOrganizationSid, destinationOrganizationSid);
+
+            }
+        }
+        return numberfound;
+    }
+
+    /**
+     * Used to order a collection by the size of PhoneNumber String
+     */
+    class NumberLengthComparator implements Comparator<IncomingPhoneNumber> {
+
+        @Override
+        public int compare(IncomingPhoneNumber o1, IncomingPhoneNumber o2) {
+            //put o2 first to make longest first in coll
+            return Integer.compare(o2.getPhoneNumber().length(), o1.getPhoneNumber().length());
+        }
+
+    }
+
+    /**
+     * This will take the regexes available in given organization, and evalute
+     * them agsint the given list of numbers, returning the first match.
+     *
+     * The list of regexes will be ordered by length to ensure the longest regexes
+     * matching any number in the list is returned first.
+     *
+     * In this case, organization details are required.
+     *
+     * @param numberQueries
+     * @param sourceOrganizationSid
+     * @param destOrg
+     * @return the longest regex matching any number in the list, null if no match
+     */
+    private IncomingPhoneNumber findByRegex(List<String> numberQueries,
+            Sid sourceOrganizationSid, Sid destOrg) {
```

Comment:
Parameter `sourceOrganizationSid` not used in the scope.

Reply:
agree, we may remove this, let me quickly fix that